crates/sui-framework/packages/bridge/sources/bridge.move

```diff
@@ -0,0 +1,349 @@
+            bridge_version: CURRENT_VERSION,
+            chain_id,
+            sequence_nums: vec_map::empty<u8, u64>(),
+            committee: committee::create(ctx),
+            treasury: treasury::create(ctx),
+            bridge_records: linked_table::new<BridgeMessageKey, BridgeRecord>(ctx),
+            frozen: false,
+        };
+        let bridge = Bridge {
+            id,
+            inner: versioned::create(CURRENT_VERSION, bridge_inner, ctx)
+        };
+        transfer::share_object(bridge);
+    }
+
+    // Create bridge request to send token to other chain, the request will be in pending state until approved
+    public fun send_token<T>(
+        self: &mut Bridge,
+        target_chain: u8,
+        target_address: vector<u8>,
+        token: Coin<T>,
+        ctx: &mut TxContext
+    ) {
+        let inner = load_inner_mut(self);
+        assert!(chain_ids::is_valid_route(inner.chain_id, target_chain), EInvalidBridgeRoute);
+        assert!(!inner.frozen, EBridgeUnavailable);
+        let bridge_seq_num = next_seq_num(inner, message_types::token());
+        let token_id = treasury::token_id<T>();
+        let token_amount = balance::value(coin::balance(&token));
+
+        // create bridge message
+        let message = message::create_token_bridge_message(
+            inner.chain_id,
+            bridge_seq_num,
+            address::to_bytes(tx_context::sender(ctx)),
+            target_chain,
+            target_address,
+            token_id,
+            token_amount,
+        );
+
+        // burn / escrow token, unsupported coins will fail in this step
+        treasury::burn(&mut inner.treasury, token, ctx);
+
+        // Store pending bridge request
+        let key = message::key(&message);
+        linked_table::push_back(&mut inner.bridge_records, key, BridgeRecord {
+            message,
+            verified_signatures: none(),
+            claimed: false,
+        });
+
+        // emit event
+        emit(TokenBridgeEvent {
+            message_type: message_types::token(),
+            seq_num: bridge_seq_num,
+            source_chain: inner.chain_id,
+            sender_address: address::to_bytes(tx_context::sender(ctx)),
+            target_chain,
+            target_address,
+            token_type: token_id,
+            amount: token_amount,
+        });
+    }
+
+    // Record bridge message approvals in Sui, called by the bridge client
+    // If already approved, return early instead of aborting.
+    public fun approve_bridge_message(
+        self: &mut Bridge,
+        message: BridgeMessage,
+        signatures: vector<vector<u8>>,
+    ) {
+        let inner = load_inner_mut(self);
+        let key = message::key(&message);
+
+        // retrieve pending message if source chain is Sui, the initial message must exist on chain.
+        if (message::message_type(&message) == message_types::token() && message::source_chain(&message) == inner.chain_id) {
```

Comment:
If the `source_chain` from `message` is not == to `inner.chain_id`, do we still want to add it in the linked_table? (as it'll fall into the `else` case).

Reply:
we want to store all approved messages in Sui as a source of truth, so as long as the signatures checks out, we will store them, also because not all messages type have a destination, we can't checks "either the source or destination.. " at this stage. If somehow someone created a message which neither destination nor source eq to `inner.chain_id`, they won't be able to claim it because we checks the route in claim token.